catalog/spatial/registry/registry-api-impl/src/main/java/org/codice/ddf/registry/api/impl/RegistryStoreCleanupHandler.java

```diff
@@ -0,0 +1,157 @@
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.codice.ddf.registry.api.internal.RegistryStore;
+import org.codice.ddf.registry.federationadmin.service.internal.FederationAdminService;
+import org.codice.ddf.security.common.Security;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.Constants;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceEvent;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.event.Event;
+import org.osgi.service.event.EventConstants;
+import org.osgi.service.event.EventHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.data.Metacard;
+
+/**
+ * This handler cleans up registry entries in the local catalog when a remote registry connection
+ * is removed.
+ */
+public class RegistryStoreCleanupHandler implements EventHandler {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RegistryStoreCleanupHandler.class);
+
+    private FederationAdminService federationAdminService;
+
+    private ExecutorService executor;
+
+    private static final int SHUTDOWN_TIMEOUT_SECONDS = 60;
+
+    private boolean cleanupRelatedMetacards = true;
+
+    private Map<Object, RegistryStore> registryStorePidToServiceMap = new ConcurrentHashMap<>();
+
+    public void destroy() {
+        executor.shutdown();
+        try {
+            if (!executor.awaitTermination(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
+                executor.shutdownNow();
+                if (!executor.awaitTermination(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
+                    LOGGER.debug("Thread pool failed to terminate");
+                }
+            }
+        } catch (InterruptedException e) {
+            executor.shutdownNow();
+        }
+    }
+
+    public void bindRegistryStore(ServiceReference serviceReference) {
+        BundleContext bundleContext = getBundleContext();
+
+        if (serviceReference != null && bundleContext != null) {
+            RegistryStore registryStore =
+                    (RegistryStore) bundleContext.getService(serviceReference);
+
+            registryStorePidToServiceMap.put(serviceReference.getProperty(Constants.SERVICE_PID),
+                    registryStore);
+        }
+    }
+
+    public void unbindRegistryStore(ServiceReference serviceReference) {
+        if (serviceReference != null) {
+            registryStorePidToServiceMap.remove(serviceReference.getProperty(Constants.SERVICE_PID));
+        }
+    }
+
+    @Override
+    public void handleEvent(Event event) {
+        if (!cleanupRelatedMetacards) {
+            return;
+        }
+
+        if (((ServiceEvent) event.getProperty(EventConstants.EVENT)).getType()
```

Comment:
Should we also make sure that `event.getProperty()` doesn't return `null` and is of the right type? I know that should not happen since we're specifically listening on the `ServiceEvent/UNREGISTERING` topic, but if someone ever messes up the blueprint then this code will start throwing NPE or CCE.

Another option would be to add an itest to verify that the unregistering works, so if someone ever messes up the blueprint we will know right away.